Add ability to configure the queue capacity for ChunkedOutput

rutterpaul-personal opened the pull request:
Adding constructors which set the queue capacity.

This allows for calling `ChunkedOutput.write` from multiple concurrent threads until the queue is full. Subsequent `write` calls will block until the queue is emptied by `flushQueue`. This change allows to prevent memory issues in case of slow clients. Basically, a backpressure mechanism based on the queue size.

This change provides a way to call write on `ChunkedOutput` from multiple threads to improve throughput, while preventing memory issues in case of slow clients. The queue is bounded and thus cannot grow endlessly.

The code now uses `put` (https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/LinkedBlockingQueue.html#put-E-) instead of `add` (https://docs.oracle.com/javase/8/docs/api/java/util/AbstractQueue.html#add-E-) so it blocks when the queue is full.

rutterpaul-personal committed:
Update ChunkedOutput.java

Adding constructors which sets the queue capacity.

This allows for calling write from different threads until the queue is full. Subsequent write calls will block until the queue is emptied.
This change allows to prevent memory issues in case of slow clients. Basically, a backpressure mechanism based on the queue size.

This change provides a way to call write on ChunkedOutput from multiple threads to improve throughput, while preventing memory issues in case of slow clients. The queue is bounded and thus cannot grow endlessly.

rutterpaul-personal committed:
Update ChunkedOutput.java

Update year

jansupol commented:
@rutterpaul-personal @paulrutter This PR just adds 6 new constructors. Would it be better to have a builder instead?

paulrutter commented:
@jansupol 
Yes, the number of constructors is not ideal. How would you see a builder pattern for this?
Keeping the current constructors as is and instead add a builder to `ChunkedOutput` which takes all configuration options?

* queueCapacity
* chunkType
* chunkDelimiter
* asyncContextProvider

That could work for me.
What do you think of adding a configuration for the queue size though?

We recently ran into memory issues because we no longer synchronized on calling the `ChunkedOutput.write` method.
This change would allow to still do so, but with backpressure in place due to the `queue.put` blocking until it's drained.

jansupol commented:
Yes, I'd keep the constructors as they are, for the backward compatibility, without adding any new (except for the one accepting the builder, setting up all the values set on the builder to the ChunkedOutput).

paulrutter commented:
Ok, i will work on that instead.

paulrutter commented:
Something like this then?
```
final ChunkedOutput<String> output = ChunkedOutput.<String>newBuilder(String.class).queueCapacity(10).chunkDelimiter("\r\n".getBytes()).build();
```

I cannot get the `chunkType` into the builder properly, as i cannot call `super` conditionally. Any ideas on that?
I now defaulted to `String.class` when it's null, which seems a safe default. I'm not even sure if we still need to set the chunkType, as the example above makes sure that `String` is used.

@jansupol what do you think about the chunkType? Can i drop it from the builder? 

I noticed that the `chunkType` is now always required in the public constructors, so i made it a required argument to the `newBuilder` method. That solves the issue.

rutterpaul-personal committed:
Update ChunkedOutput.java

Use builder instead.

rutterpaul-personal committed:
Update ChunkedOutput.java

rutterpaul-personal committed:
Update ChunkedOutput.java

Copyright

rutterpaul-personal committed:
Update ChunkedOutput.java

Add chunkType, String by default.

rutterpaul-personal committed:
Update ChunkedOutput.java

Add javadoc

rutterpaul-personal committed:
Update ChunkedOutput.java

Typo

rutterpaul-personal committed:
Update ChunkedOutput.java

Hide constructor

paulrutter commented:
If these changes are ok, documentation probably needs updating on https://eclipse-ee4j.github.io/jersey.github.io/documentation/2.42/async.html#chunked-output and it would need a unit test to test the builder.

rutterpaul-personal committed:
Update ChunkedInputOutputTest.java

Add e2e test that uses builder with queue capacity of 2.

rutterpaul-personal committed:
Update ChunkedInputOutputTest.java

copyright

rutterpaul-personal committed:
Update async.xml

Update documentation with an example of the builder with the capacity

rutterpaul-personal committed:
Update async.xml

Copyright

rutterpaul-personal committed:
Update ChunkedOutput.java

Make chunkType a required argument to the builder.
Change builder methods.

rutterpaul-personal committed:
Update async.xml

Update doc.

rutterpaul-personal committed:
Update ChunkedInputOutputTest.java

Update test

rutterpaul-personal committed:
Update ChunkedInputOutputTest.java

Fix checkstyle issue

jansupol commented:
The builder looks good. I'd prefer the name `builder` over `newBuilder`, though.

The chunkType is required to know the type in runtime, or the type can be known from the ChunkedOutput subclass (hence the protected constructors), the same way the GenericType works. 

So, with the protected constructor functionality, If needed, I think the builder could work with constructs like: 

```
new ChunkedOutput<List<String>>(ChunkedOutput.builder(size).build()) { }

protected ChunkedOutput(Builder builder) {
  super(builder.getChunkedType() != null ? builder.getChunkedType() : getType());
  ...
}
```
wdyt?

paulrutter commented:
@jansupol That would make sense, except that i cannot call `getType` in the constructor before `super` has been called

`Cannot reference 'GenericType.getType()' before supertype constructor has been called`

Or do i misunderstand?

jansupol commented:
@paulrutter My bad, it's not that simple. Would need to have two constructors for that... And two builders... Maybe something like

```
class TypedBuilder extends Builder {
   private Type chunkedType; // move here from Builder
   TypedBuilder(Type type){...}
   ChunkedOutput build() {return new ChunkedOutput(this);} // perhaps throwing a warning in Builder
}

static TypedBuilder builder(Type type) { return new TypedBuilder(type());}
static Builder builder(){return new Builder());

protected ChunkedOutput(Builder) {  super(); .... }
private ChunkedOutput(TypedBuilder builder) { super(builder.getType()); ... }
```
Perhaps too complicated at the moment ... unless you would have a use case for it.

rutterpaul-personal committed:
Update ChunkedOutput.java

Add two builders, one with and one without type.

rutterpaul-personal committed:
Update ChunkedInputOutputTest.java

Update test

rutterpaul-personal committed:
Update async.xml

Update docs

paulrutter commented:
@jansupol i made the changes, could you check?
Not sure where we would put a warning though.

rutterpaul-personal committed:
Update ChunkedInputOutputTest.java

Copyright

rutterpaul-personal committed:
Update ChunkedOutput.java

jansupol commented on tests/e2e-server/src/test/java/org/glassfish/jersey/tests/e2e/server/ChunkedInputOutputTest.java:
```diff
@@ -60,14 +60,33 @@ public class ChunkedInputOutputTest extends JerseyTest {
+        @GET
+        @Path("/testWithBuilder")
+        public ChunkedOutput<String> getWithBuilder() {
+            return getOutput(ChunkedOutput.<String>builder(String.class).queueCapacity(2).chunkDelimiter("\r\n".getBytes()).build());
```
This is likely the line where the checkstyle fails as the maximum row length is set to 130

paulrutter replied:
I will correct it in a sec

paulrutter replied:
Sorry for all the separate commits, having some trouble having two Github accounts atm.

rutterpaul-personal committed:
Update ChunkedInputOutputTest.java

Fix checkstyle

paulrutter commented:
I would also need to check if a potential deadlock can occur (a collegue pointed it out to me), when an exception would occur on flushing the queue.

rutterpaul-personal committed:
Update ChunkedOutput.java

Drain queue on exception, to prevent deadlock in threads waiting on the queue to be drained while an exception occurred.

paulrutter commented:
@jansupol all checks succeeded. Can you check the changes when you have time?
I added code which clears the queue on an exception, to prevent potential deadlock.

Further, all proposed changes are present, including the two builders.

jansupol reviewed (commented):
LGTM, thanks

paulrutter commented:
Any idea in which version this will land?

jansupol commented:
2.43 the next week hopefully

paulrutter commented:
I assume it will also land in 3.x?

jansupol commented:
> I assume it will also land in 3.x?

Sure, 3.0 right after 2.43, 3.1 a bit later, 4.0 next.

paulrutter commented:
Confirmed it works on 2.43.
